cql3/statements/select_statement.cc

```diff
@@ -1341,7 +1344,24 @@ indexed_table_select_statement::find_index_clustering_rows(query_processor& qp,
                 return row.get_blob(cdef.name_as_text());
             });
             auto ck = clustering_key::from_range(ck_columns);
+
+            if (_index.target_type() == cql3::statements::index_target::target_type::collection_values) {
+                // The index on collection values is special in a way, as its' clustering key contains not only the
+                // base primary key, but also a column that holds the keys of the cells in the collection, which
+                // allows to distinguish cells with different keys but the same value.
+                // This has an unwanted consequence, that it's possible to receive two identical base table primary
+                // keys. Thankfully, they are guaranteed to occur consequently.
+                if (last_primary_key) {
+                    const auto& [last_dk, last_ck] = last_primary_key->get();
+                    if (last_dk.equal(*_schema, dk) && last_ck.equal(*_schema, ck)) {
+                        continue;
+                    }
+                }
+            }
+
```

Comment:
What happens when paging, if a page boundary splits this check?

Reply:
Good question. I'll write a test for this. If it doesn't work (and I'm worried it doesn't), I don't know how to fix it. But I'll cross that bridge when I come to it (after writing a test).

Reply:
@avikivity was absolutely right, this is a bug, and I confirmed it exists in a test. I'll push the xfailing test to the ever-growing series, but then need to think whether to solve this now, and how.

Reply:
Pushed patch 32 with a test for this bug.

Reply:
Pushed patch 33 ("secondary index: fix paging in map value indexing") fixing this bug and making the tests for it pass.

Reply:
I made the test stronger and found a variant of this bug still exists in the special case where the base table has only partition keys. So need to fix that too :-( What a mess :-(

Reply:
I know where to fix the case of only partition keys (there's separate code for that case), but it's harder to fix. This whole indexing code is so complicated and so interconnected :-(

Reply:
I have good news and bad news. The good news is that I actually fixed the case for partition keys correctly.

The bad news is that in both partition-key-only and partition-and-clustering-key cases, if there are any **empty pages** in the middle of the response - because a page of one item was a duplicate and discarded - the query stops prematurely. Only my partition-key test had this case which is why I thought the new bug was specific to partition keys.

Reply:
It turns out that I fell in exactly the **driver bug** about which I told @denesb about today - older versions of the Cassandra/Scylla drivers stop reading when reaching an empty page, which is a bug and was fixed later. I have a fixture `driver_bug_1` which skips the test if the driver is too old, and now I added it and the failing tests got skipped. 

Now I need to update the driver to see if these tests now pass.

Reply:
Ok, good news: I pushed a new patch 33 which makes the tests stronger and passes all the tests. I had to fix the partition-key-only and partition-and-clustering-key code in two separate places, but using exactly the same technique and it's fairly straightforward. So I think I'm done with this TODO as well, and perhaps this series is finally ready for prime time?